bon_csv2html.1
.TH bon_csv2html 1 
.SH "NAME"
.BR bon_csv2html
program to convert CSV format Bonnie++ data to a HTML form
using tables suitable for display on a web page. NB Lynx can't display this
properly, and due to the size it probably requires 1024x768 monitor to display
properly.
.P
.BR bon_csv2txt
program to convert CSV format Bonnie++ data to plain-text format
suitable for pasting into an email or reading on a Braille display.

.SH "DESCRIPTION"
They take CSV format (comma-delimited spreadsheet files AKA Comma Seperated
Values in MS land) data on standard input and produce HTML or plain text on
standard output respectively.

.SH "FORMAT"
This is a list of the fields used in the CSV files format version 2.  Format
version 1 was the type used in Bonnie++ < 1.90.
Before each field I list the field number as well as the name given in the heading
.TP
.B 0 format_version
Version of the output format in use (1.93b)
.TP
.B 1 bonnie_version
(1.93b)
.TP
.B 2 name
Machine Name
.TP
.B 3 concurrency
The number of copies of each operation to be run at the same time
.TP
.B 4 seed
Random number seed
.TP
.B 5 file_size
Size in megs for the IO tests
.TP
.B 6 chunk_size
Size of chunks in bytes
.TP
.B 7 putc,putc_cpu
Results for writing a character at a time K/s,%CPU
.TP
.B 9 put_block,put_block_cpu
Results for writing a block at a time K/s,%CPU
.TP
.B 11 rewrite,rewrite_cpu
Results for reading and re-writing a block at a time K/s,%CPU
.TP
.B 13 getc,getc_cpu
Results for reading a character at a time K/s,%CPU
.TP
.B 15 get_block,get_block_cpu
Results for reading a block at a time K/s,%CPU
.TP
.B 17 seeks,seeks_cpu
Results for the seek test seeks/s,%CPU
.TP
.B 19 num_files
Number of files for file-creation tests (units of 1024 files)
.TP
.B 20 max_size
The maximum size of files for file-creation tests.  Or the type of files for
links.
.TP
.B 21 min_size
The minimum size of files for file-creation tests.
.TP
.B 22 num_dirs
The number of directories for creation of files in multiple directories.
.TP
.B 23 file_chunk_size
The size of blocks for writing multiple files.
.TP
.B 24 seq_create,seq_create_cpu
Rate of creating files sequentially files/s,%CPU
.TP
.B 26 seq_stat,seq_stat_cpu
Rate of reading/stating files sequentially files/s,%CPU
.TP
.B 28 seq_del,seq_del_cpu
Rate of deleting files sequentially files/s,%CPU
.TP
.B 30 ran_create,ran_create_cpu
Rate of creating files in random order files/s,%CPU
.TP
.B 32 ran_stat,ran_stat_cpu
Rate of deleting files in random order files/s,%CPU
.TP
.B 34 ran_del,ran_del_cpu
Rate of deleting files in random order files/s,%CPU
.TP
.B 36 putc_latency,put_block_latency,rewrite_latency
Latency (maximum amount of time for a single operation) for putc, put_block,
and reqrite
.TP
.B 39 getc_latency,get_block_latency,seeks_latency
Latency for getc, get_block, and seeks
.TP
.B 42 seq_create_latency,seq_stat_latency,seq_del_latency
Latency for seq_create, seq_stat, and seq_del
.TP
.B 45 ran_create_latency,ran_stat_latency,ran_del_latency
Latency for ran_create, ran_stat, and ran_del
.P
A string that starts with '#' is a comment.

.SH "AUTHOR"
These programs were written by Russell Coker <russell@coker.com.au>. May be
freely used and distributed without restriction.


debian/changelog
bonnie++ (1.00) unstable; urgency=low

  * License is now GPL.

  * Now include ZCAV in the same package.  ZCAV package should disappear.

  * ZCAV reads through a hard drive sequentially and reports the IO speeds for
    different zones of the drive.

  * Fixed a few minor issues with the documentation, and put the test programs
    in /usr/sbin as they aren't generally run by regular users.  Also use man
    section 8.

 -- Russell Coker <russell@coker.com.au>  Sat, 01 Mar 2000 12:01:00 +0100

bonnie++ (0.99j) unstable; urgency=low

  * 0.99h core dumped when you didn't specify "-b" for file creation tests,
    fixed.

 -- Russell Coker <russell@coker.com.au>  Sun, 05 Mar 2000 11:16:42 +0100

bonnie++ (0.99h) unstable; urgency=low

  * Fixed a variety of bugs in the semaphore code which were introduced in
    0.99g.

  * Fixed formatting of output.

  * Added "-b" option to sync all writes.

 -- Russell Coker <russell@coker.com.au>  Sun, 05 Mar 2000 11:16:42 +0100

bonnie++ (0.99g) unstable; urgency=low

  * Now use getopt() for checking command-line options.

  * Added new versions of fork and semaphore code, initially developed for
    postal.

  * Fixed the message that's displayed when bad command-line parameters are
    entered.

 -- Russell Coker <russell@coker.com.au>  Wed, 23 Feb 2000 22:16:23 +0100

bonnie++ (0.99f) unstable; urgency=low

  * Added "-f" parameter to skip per-char tests and semaphore code to
    synchronise multiple instances of Bonnie++.  Thanks to
    Christian Kagerhuber <c.kagerhuber@t-online.net> for the patch!

  * Added srand() after the fork so each child gets different random numbers.

 -- Russell Coker <russell@coker.com.au>  Wed, 12 Jan 2000 16:45:28 +1100

bonnie++ (0.99e) unstable; urgency=low

  * Fixed the operation of "-x" parameter (used to just cause crashes).
 
  * Made it cleanly exit under some error conditions where it used to crash.

  * Merged code from the OS/2 port, needs lots of testing...

 -- Russell Coker <russell@coker.com.au>  Wed, 12 Jan 2000 16:45:28 +1100

bonnie++ (0.99d) unstable; urgency=low

  * Added some more functionality. Tests hard and soft link creation.

 -- Russell Coker <russell@coker.com.au>  Sun, 21 Nov 1999 22:37:42 +0200

bonnie++ (0.99c) unstable; urgency=low

  * Fix some bugs with big IO and include man pages.

 -- Russell Coker <russell@coker.com.au>  Sun, 21 Nov 1999 22:37:42 +0200

bonnie++ (0.99b) unstable; urgency=low

  * Initial Release.

 -- Russell Coker <russell@coker.com.au>  Wed, 13 Oct 1999 22:15:53 +0200

Local variables:
mode: debian-changelog
End:


bon_csv2html.cpp
#include "bonnie.h"
#include <stdio.h>
#include <vector>
#include <string.h>
#include <math.h>

#define MAX_ITEMS 33
#ifdef WIN32
using namespace std;
#endif
typedef vector<PCCHAR> STR_VEC;

vector<STR_VEC *> data;
typedef PCCHAR * PPCCHAR;
PPCCHAR * props;

void header();
void read_in(CPCCHAR buf);
void print_a_line(int num, int start, int end);
void footer();
void calc_vals();
PCCHAR get_col(double range_col, double val);
const int vals[MAX_ITEMS] =
  { 0,0,0,0,0,1,2,1,2,1,2,1,2,1,2,1,2,
    0,0,0,0,1,2,1,2,1,2,1,2,1,2,1,2 };

void usage()
{
  exit(1);
}

int main(int argc, char **argv)
{
  header();

  char buf[1024];

  FILE *fp = NULL;
  if(argc > 1)
  {
    fp = fopen(argv[1], "r");
    if(!fp)
      usage();
  }
  while(fgets(buf, sizeof(buf), fp ? fp : stdin))
  {
    buf[sizeof(buf) - 1] = '\0';
    strtok(buf, "\r\n");
    read_in(buf);
  }
//  $vals{putc} = 5;
//  $vals{put_block} = 7;
//  $vals{rewrite} = 9;
//  $vals{getc} = 11;
//  $vals{get_block} = 13;
//  $vals{seeks} = 15;
//  $vals{seq_create} = 21;
//  $vals{seq_stat} = 23;
//  $vals{seq_del} = 25;
//  $vals{ran_create} = 27;
//  $vals{ran_stat} = 29;
//  $vals{ran_del} = 31;

  props = new PPCCHAR[data.size()];
  unsigned int i;
  for(i = 0; i < data.size(); i++)
  {
    props[i] = new PCCHAR[MAX_ITEMS];
    props[i][0] = NULL;
    props[i][1] = NULL;
    props[i][2] = "bgcolor=\"#FFFFFF\" class=\"rowheader\"><FONT SIZE=+1";
    props[i][3] = "class=\"size\" bgcolor=\"#FFFFFF\"";
    int j;
    for(j = 4; j < MAX_ITEMS; j++)
    {
      if( (j >= 17 && j <= 20) || j == 4 )
      {
        props[i][j] = "class=\"size\" bgcolor=\"#FFFFFF\"";
      }
      else
      {
//        props[i][j] = "bgcolor=\"#FFFFFF\"";
        props[i][j] = NULL;
      }
    }
  }
  if(data.size() > 1)
  {
    calc_vals();
  }
  for(i = 0; i < data.size(); i++)
  {
    print_a_line(i, 2, 32);
  }
  footer();
  return 0;
}

typedef struct { double val; int pos; int col_ind; } ITEM;
typedef ITEM * PITEM;

int compar(const void *a, const void *b)
{
  double a1 = PITEM(a)->val;
  double b1 = PITEM(b)->val;
  if(a1 < b1)
    return -1;
  if(a1 > b1)
    return 1;
  return 0;
}

void calc_vals()
{
  ITEM *arr = new ITEM[data.size()];
  for(unsigned int column_ind = 0; column_ind < MAX_ITEMS; column_ind++)
  {
    if(vals[column_ind] == 1)
    {
      unsigned int row_ind;
      for(row_ind = 0; row_ind < data.size(); row_ind++)
      {
        if((*data[row_ind]).size() <= column_ind
         || sscanf((*data[row_ind])[column_ind], "%lf", &arr[row_ind].val) == 0)
          arr[row_ind].val = 0.0;
        arr[row_ind].pos = row_ind;
      }
      qsort(arr, data.size(), sizeof(ITEM), compar);
      int ind = -1;
      double min_col = -1.0, max_col = -1.0;
      for(row_ind = 0; row_ind < data.size(); row_ind++)
      {
        // if item is different from previous then increment col index
        if(row_ind == 0 || arr[row_ind].val != arr[row_ind - 1].val)
        {
          if(arr[row_ind].val != 0.0)
          {
            ind++;
            if(min_col == -1.0)
              min_col = arr[row_ind].val;
            else
              min_col = __min(arr[row_ind].val, min_col);
            max_col = __max(max_col, arr[row_ind].val);
          }
        }
        arr[row_ind].col_ind = ind;
      }
      if(ind > 0)
      {
        double divisor = max_col / min_col;
        if(divisor < 2.0)
        {
          double mult = sqrt(2.0 / divisor);
          max_col *= mult;
          min_col /= mult;
        }
        double range_col = max_col - min_col;
        for(row_ind = 0; row_ind < data.size(); row_ind++)
        {
          if(arr[row_ind].col_ind > -1)
            props[row_ind][column_ind]
                          = get_col(range_col, arr[row_ind].val - min_col);
        }
      }
    }
  }
}

PCCHAR get_col(double range_col, double val)
{
  const int buf_len = 18;
  PCHAR buf = new char[buf_len];
  int green = int(255.0 * val / range_col);
  green = __min(green, 255);
  int red = 255 - green;
  _snprintf(buf, buf_len, "bgcolor=\"#%02X%02X00\"", green, red);
  buf[buf_len - 1] = '\0';
  return buf;
}

void heading(const char * const head);

void header()
{
  printf("<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.0//EN\" \"http://www.w3.org/TR/REC-html40/strict.dtd\">"
"<HTML>"
"<HEAD><TITLE>Bonnie++ Benchmark results</TITLE>"
"<STYLE type=\"text/css\">"
"TD.header {text-align: center; backgroundcolor: \"#CCFFFF\" }"
"TD.rowheader {text-align: center; backgroundcolor: \"#CCCFFF\" }"
"TD.size {text-align: center; backgroundcolor: \"#CCCFFF\" }"
"TD.ksec {text-align: center; fontstyle: italic }"
"</STYLE>"
"<BODY>"
"<TABLE ALIGN=center BORDER=3 CELLPADDING=2 CELLSPACING=1>"
"<TR><TD COLSPAN=3 class=\"header\"><FONT SIZE=+1><B>"
"Version " BON_VERSION
"</B></FONT></TD>"
"<TD COLSPAN=6 class=\"header\"><FONT SIZE=+2><B>Sequential Output</B></FONT></TD>"
"<TD COLSPAN=4 class=\"header\"><FONT SIZE=+2><B>Sequential Input</B></FONT></TD>"
"<TD COLSPAN=2 ROWSPAN=2 class=\"header\"><FONT SIZE=+2><B>Random<BR>Seeks</B></FONT></TD>"
"<TD COLSPAN=4 class=\"header\"></TD>"
"<TD COLSPAN=6 class=\"header\"><FONT SIZE=+2><B>Sequential Create</B></FONT></TD>"
"<TD COLSPAN=6 class=\"header\"><FONT SIZE=+2><B>Random Create</B></FONT></TD>"
"</TR>"
"<TR><TD></TD>"
"<TD>Size</TD><TD>Chunk Size</TD>");
  heading("Per Char"); heading("Block"); heading("Rewrite");
  heading("Per Char"); heading("Block");
  printf("<TD>Num Files</TD><TD>Max Size</TD><TD>Min Size</TD><TD>Num Dirs</TD>");
  heading("Create"); heading("Read"); heading("Delete");
  heading("Create"); heading("Read"); heading("Delete");
  printf("</TR>");

  printf("<TR><TD COLSPAN=3></TD>");

  int i;
  CPCCHAR ksec_form = "<TD class=\"ksec\"><FONT SIZE=-2>%s/sec</FONT></TD>"
                      "<TD class=\"ksec\"><FONT SIZE=-2>%% CPU</FONT></TD>";
  for(i = 0; i < 5; i++)
  {
    printf(ksec_form, "K");
  }
  printf(ksec_form, "");
  printf("<TD COLSPAN=4></TD>");
  for(i = 0; i < 6; i++)
  {
    printf(ksec_form, "");
  }
  printf("</TR>\n");
}

void heading(const char * const head)
{
  printf("<TD COLSPAN=2>%s</TD>", head);
}

void footer()
{
  printf("</TABLE>\n</BODY></HTML>\n");
}

STR_VEC *split(CPCCHAR delim, CPCCHAR buf)
{
  STR_VEC *arr = new STR_VEC;
  char *tmp = strdup(buf);
  while(1)
  {
    arr->push_back(tmp);
    tmp = strstr(tmp, delim);
    if(!tmp)
      break;
    *tmp = '\0';
    tmp += strlen(delim);
  }
  return arr;
}

void read_in(CPCCHAR buf)
{
  STR_VEC *arr = split(",", buf);
  if(strcmp((*arr)[0], "2") )
  {
    fprintf(stderr, "Can't process: %s\n", buf);
    free((void *)(*arr)[0]);
    delete arr;
    return;
  }

//printf("str:%s,%s,%s,%s\n", (*arr)[2], (*arr)[3], (*arr)[4], (*arr)[5]);
  data.push_back(arr);
}

void print_a_line(int num, int start, int end)
{
  STR_VEC *arr = data[num];

  printf("<TR>");
  int i;
  for(i = start; i < end; i++)
  {
    PCCHAR line_data = (*arr)[i];
    if(!line_data) line_data = "";
    if(props[num][i])
      printf("<TD %s>%s</TD>", props[num][i], line_data);
    else
      printf("<TD>%s</TD>", line_data);
  }
  printf("</TR>\n");
}
